### PR TITLE
Firewall State Automation Implementation

### PR DESCRIPTION
**Pull request by TravisFallsMSFT**
# Pull Request 

- [x] Documentation changes
- [x] Code changes
- [ ] Test changes
- [ ] CI-CD changes
- [ ] GitHub Template changes

## PR Checklist

- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] My code follows the code style of this project.
- [x] I ran lint checks locally prior to submission.
- [x] Have you checked to ensure there aren't other open Pull Requests for the same update/change?

## Purpose of PR

This Pull Request implements changes that will save on Azure expense for Microsoft Corp. This runbook will enable and disable the Azure firewall as well as the metric and log-based alert rules associated with the firewall.  

## Implementation 

1. Establishes documentation that illustrates the steps to implement a user-assigned managed identity-based runbook.  
2. Implement the Azure Automation Account.
3. Implement an Azure user-assigned managed identity.
4. Assign the Firewall role to the user-assigned managed identity.
5. Implement a PowerShell-based Runbook that starts and stops the Azure Firewall as well as it disables and enables Log, and Metric-base alert rules. 
6. Implement an Automation schedule for the runbook.

## Does this introduce a breaking change

- [ ] YES
- [x] NO

## Validation

- [ ] Unit tests updated and ran successfully
- [x] Update documentation or issue referenced above

## Issues Closed or Referenced

- Closes #1031
- References #1030, #1003, #815

### CODE REVIEW

File: docs/deployAutomationForFirewallCostOptimization.md [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+   ```bash
+
+    # Log into the tenant
+    az login --tenant {tenant Id}
+
+    # Set the subscription desired    
+    az account set --subscription {subscription name}
+    
+    # pass in resource group  name and location
+    az group create -name {automation resource group name} -location {location}
```
Comment (AAkindele) [inline docs/deployAutomationForFirewallCostOptimization.md:206]: Replace the `{tenant Id}` style replacements with variables. That way, it is easier to copy and paste.
The top of each code block could have all the variables needed in one spot for easy replacement.

ex:
```suggestion

    ASB_TENANT_ID={tenant Id}
    ASB_SUB_NAME={subscription name}
    ASB_FW_AUTO_LOCATION={location}
    ASB_FW_AUTO_RG={automation resource group name}

    # Log into the tenant
    az login --tenant $ASB_TENANT_ID

    # Set the subscription desired    
    az account set --subscription $ASB_SUB_NAME
    
    # pass in resource group  name and location
    az group create -name $ASB_FW_AUTO_RG -location $ASB_FW_LOCATION
```

File: docs/deployAutomationForFirewallCostOptimization.md [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+
+# Log into the tenant
+az login --tenant {tenant Id}
+
+# Set the subscription desired    
+az account set --subscription {subscription name}
+
+az identity create --resource-group {automation resource group name} --name {user-assigned managed identity name}
+
+# The Az CLI commands for this errored at the time of writing so pwsh was used.
```
Comment (AAkindele) [inline docs/deployAutomationForFirewallCostOptimization.md:224]: If there is no task for this yet, I recommend creating a new one error so we can investigate further if needed.

File: docs/deployAutomationForFirewallCostOptimization.md [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+
+1. LInk the Runbook to the schedules
+
+### ADVANCED - Automated Scripts to Run
+
+BEFORE continuing please make sure all requirements have been met in the section labeled [prerequisites]("#-prerequisites").
+
+1. [Create Automation Infrastructure (BASH script)]("./scripts/automatoin/createAutomationForFirewallCostOptimization.sh")
+  
+  The [createAutomationForFirewallCostOptimization.sh]("./scripts/automation/createautimationForFirewallCostOptimization.sh") script "dot sources" the [firewallAutomationForCostOptimization.variables.sh]("./scripts/automation/cfirewallAutomationForCostOptimization.variables.sh").
```
Comment (AAkindele) [inline docs/deployAutomationForFirewallCostOptimization.md:353]: Thoughts on making use of the saveenv.sh script and reusing some of variable defined in the main README? For example, there is already `ASB_ENV`, `ASB_SUBSCRIPTION_ID`, `ASB_TENANT_ID`, etc. The benefit there is that all the environment variables will end up in the same environment specific file. ex: `ngsa-asb-dev-eastus.env`

Comment (TravisFallsMSFT) [reply docs/deployAutomationForFirewallCostOptimization.md:353]: I was struggling to have PowerShell read the envs that were set in BASH and vice versa, which is why I simplified.  I am glad to take another look though.

File: docs/deployAutomationForFirewallCostOptimization.md [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+   2. Azure Automation Account
+   3. Azure PowerShell Runbook
+  
+<img src="./assets/automation/listOfResourcesInResourceGroup.png" alt="List of resources that will be created when complete." />
+
+## Choosing the Installation Method
+
+The instructions below show three manners to accomplish the same objective.  
+
+Choose **ONLY** one.
```
[the remark this answers is not in the document]
Comment (TravisFallsMSFT) [reply docs/deployAutomationForFirewallCostOptimization.md:87]: I have removed the two manual steps and the docs associated.  I have also created a firewall-automation folder within the Spikes and put the other two ways there so we don't lose the docs.

File: docs/deployAutomationForFirewallCostOptimization.md [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+2. [CLI & Az PowerShell Modules](#azure-cli-and-az-powershell-modules) - Using a terminal and executing the PowerShell and Bash CLI commands interactively
+3. Executing the Bash and PowerShell Scripts Provided.  
+
+### Portal
+
+Note if you chose a "portal based installation" please follow the instructions below.  
+
+Create a resource group within the subscription and tenant that contains the firewalls.  This new resource group will house three resources as shown below.
+
+<img src="./assets/automation/createResourceGroup1Portal.png" alt="Create a new Resource Group" />
```
[the remark this answers is not in the document]
Comment (TravisFallsMSFT) [reply docs/deployAutomationForFirewallCostOptimization.md:99]: Fixed; I had issues back a few years using these with github.  All good to use them :-)

File: docs/deployAutomationForFirewallCostOptimization.md (hunk 1 above)
Comment (TravisFallsMSFT) [reply docs/deployAutomationForFirewallCostOptimization.md:206]: completed this.

File: docs/Firewall-Automation.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+|  5. | Powershell Content in Runbook             |                               [link](https://learn.microsoft.com/en-us/powershell/module/az.automation/import-azautomationrunbook?view=azps-8.3.0)                               | Upload [pre-defined Powershell content](../automation/FirewallToggle.ps1) into the Runbook body.                                                                            |   |
+|  6. | Automation Schedule(s) _using Powershell_ |                               [link](https://learn.microsoft.com/en-us/powershell/module/az.automation/import-azautomationrunbook?view=azps-8.3.0)                               | Create the schedules that will execute the Firewall automation.  These had to be created using Powershell instead of the Azure CLI.  No equivalent behavior has been found. |   |
+
+### Resources Created When Complete
+
+1.  Azure user-assigned Managed Identity
+2.  Azure Automation Account
+3.  Azure PowerShell Runbook
+
+    ![List of resources that will be created when complete.](./assets/automation/listOfResourcesInResourceGroup.png)
```
Comment (PurpleBriar) [inline docs/Firewall-Automation.md:67]: Update this link. It's broken

File: scripts/Firewall-Automation/Firewall-Automation-Infrastructure-Variables.sh [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+export ASB_FW_Subscription_Name='jofultz-team' # Suscription Id for Jofultz-Team
+export ASB_FW_Base_NSGA_Name='ngsa-asb'
+export ASB_FW_Base_Automation_System_Name='firewall-automation'
+export ASB_FW_Environment='dev'
+export ASB_FW_PowerShell_Runbook_File_Name='Firewall-Automation-Runbook.ps1' # Powershell based runbook file name.
+export ASB_FW_Sku='Basic' # Sku for the Automation Account
+export ASB_FW_Location='westus' # Location for resource creation
+export ASB_FW_PowerShell_Runbook_Description='This runbook automates the allocation and de-allocation of a firewall for the purposes of scheduling.' # Description for the runbook.
+
+./saveenv.sh
```
Comment (PurpleBriar) [inline scripts/Firewall-Automation/Firewall-Automation-Infrastructure-Variables.sh:14]: Add new line at end of script

File: scripts/Firewall-Automation/Firewall-Automation-Infrastructure.sh [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+function UpdateAzureAutomationAccountToAllowSystemAssignedIdentity() {
+  # Arguments: 
+    # parameter position 1 = Automation Account Name
+    # parameter position 2 = Automation Resource Group
+    # parameter position 3 = Subscription Id
+    # parameter position 4 = Tenant Id
+    # parameter position 5 = User-Assigned Managed Identity Name
+  
+  echo
+  echo "Assigning role Managed Identity Operator to the System Assigned Identity for automation account ${1} in resource group ${2}, within subscription id ${3}..."
```
Comment (PurpleBriar) [inline scripts/Firewall-Automation/Firewall-Automation-Infrastructure.sh:156]: Looks like you're using the user-assigned managed identity

File: spikes/Firewall-Automation/Firewall-Automation-Manual-Installation.md [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+
+![Create a new Resource Group](./assets/automation/createResourceGroup1Portal.png)
+
+#### Create User-Assigned Managed Identity
+
+Navigate to the newly created resource group and click the "+" to add a new resource
+
+Search for "User Assigned Managed Identity".
+
+![Create a User-Assigned Managed Identity for the Automation](./assets/automation/createUserAssginedManagedIdentityPortal.png)
```
Comment (PurpleBriar) [inline spikes/Firewall-Automation/Firewall-Automation-Manual-Installation.md:107]: Please update broke links. ./assets/automation/createUserAssginedManagedIdentityPortal.png should point to ./assets/Firewall-Automation/createUserAssginedManagedIdentityPortal.png

**Review by AAkindele — APPROVED** [review]
Approved pending @PurpleBriar comments.

File: docs/Firewall-Automation.md (hunk 6 above)
Comment (TravisFallsMSFT) [reply docs/Firewall-Automation.md:67]: fixed path issue.

File: scripts/Firewall-Automation/Firewall-Automation-Infrastructure-Variables.sh (hunk 7 above)
Comment (TravisFallsMSFT) [reply scripts/Firewall-Automation/Firewall-Automation-Infrastructure-Variables.sh:14]: fixed issue.

File: spikes/Firewall-Automation/Firewall-Automation-Manual-Installation.md (hunk 9 above)
Comment (TravisFallsMSFT) [reply spikes/Firewall-Automation/Firewall-Automation-Manual-Installation.md:107]: I deleted the "spike" folder for Firewall-Automation as it wasn't really a spike but different ways to do the same task; more technical debt in my opinion.

**Review by PurpleBriar — APPROVED** [review]

**Pull request merged**